File: solutions/graphs/bfs/openTheLock.js
const START_COMBINATION = '0000';
const IMPOSSIBLE_TO_OPEN = -1;

const goUp = (char) => {
    if (char === '9') {
        const newCharInt = 0;

        return newCharInt.toString();
    }

    const charInt = parseInt(char);
    const newCharInt = charInt + 1;

    return newCharInt.toString();
};

const goDown = (char) => {
    if (char === '0') {
        const newCharInt = 9;

        return newCharInt.toString();
    }

    const charInt = parseInt(char);
    const newCharInt = charInt - 1;

    return newCharInt.toString();
};

const replaceCharAtI = (combination, i, char) => {
    const beforeI = combination.substring(0, i);
    const afterI = combination.substring(i + 1);

    return beforeI + char + afterI;
};

const getNeighbors = (combination, queue) => {
    const allNeighbors = [];

    for (let i = 0; i < combination.length; i++) {
        const char = combination[i];

        const upChar = goUp(char);
        const downChar = goDown(char);

        const upString = replaceCharAtI(combination, i, upChar);
        const downString = replaceCharAtI(combination, i, downChar);

        allNeighbors.push(upString);
        allNeighbors.push(downString);
    }

    return allNeighbors;
};

const getMinTurns = (target, deadends) => {
    const queue = new Queue();
    queue.enqueue([START_COMBINATION, 0]); // [combination, numTurnsSoFar]

    const visited = new Set();

    while (queue.size() > 0) {
        // Remove node
        const [combination, numTurnsSoFar] = queue.dequeue();

        // Process node

        // Note: we have tried putting the "deadends.has" and "visited.has" checks
        // inside of the "Add neighbors" step, but we have not been able to get
        // this to work on LeetCode. As of now, we're not sure why.
        if (deadends.has(combination)) continue;

        // This is an undirected graph. From "1111" we can go to "1112" and from
        // "1112" we can go to "1111".
        // Thus, this graph CAN have cycles (all undirected graphs can have cycles),
        // so we must tracked visited nodes to prevent infinite loops.
        if (visited.has(combination)) continue;

        if (combination == target) return numTurnsSoFar;

        visited.add(combination);

        // Add neighbors
        const neighbors = getNeighbors(combination, queue);

        for (const neighborCombination of neighbors) {
            queue.enqueue([neighborCombination, numTurnsSoFar + 1]);
        }
    }

    // If we break out of our loop (meaning we never hit our `return numTurnsSoFar`),
    // we know that it is NOT possible to reach `target`
    return IMPOSSIBLE_TO_OPEN;
};

const openLock = function (deadends, target) {
    const deadendsSet = new Set(deadends);

    const includesStart = deadendsSet.has(START_COMBINATION);
    const includesTarget = deadendsSet.has(target);

    if (includesStart || includesTarget) return IMPOSSIBLE_TO_OPEN;

    return getMinTurns(target, deadendsSet);
};

File: solutions/python/graphs/bfs/open_the_lock.py
START_COMBINATION = '0000'
IMPOSSIBLE_TO_OPEN = -1

COMBINATION = 'combination'
NUM_TURNS_SO_FAR = 'num_turns_so_far'

ZERO_STRING = '0'
NINE_STRING = '9'

class Solution:
    def go_up(self, char):
        if char == NINE_STRING:
            return ZERO_STRING
        
        char_int = int(char)
        new_char_int = char_int + 1
        
        return str(new_char_int)
    
    def go_down(self, char):
        if char == ZERO_STRING:
            return NINE_STRING
        
        char_int = int(char)
        new_char_int = char_int - 1
        
        return str(new_char_int)
    
    def replace_char_at_i(self, combination, i, char):
        before_i = combination[:i]
        after_i = combination[i + 1:]
        
        return before_i + char + after_i
    
    def get_neighbors(self, combination):
        all_neighbors = []
        
        for i, char in enumerate(combination):
            up_char = self.go_up(char)
            down_char = self.go_down(char)
            
            up_string = self.replace_char_at_i(combination, i, up_char)
            down_string = self.replace_char_at_i(combination, i, down_char)
            
            all_neighbors.append(up_string)
            all_neighbors.append(down_string)
            
        return all_neighbors
            
    def get_min_turns(self, target, deadends):
        queue = deque()
        queue.append({
            COMBINATION: START_COMBINATION,
            NUM_TURNS_SO_FAR: 0 
        })
        
        visited = set()
        
        while queue:
            # Remove node
            node = queue.popleft()
            
            combination = node[COMBINATION]
            num_turns_so_far = node[NUM_TURNS_SO_FAR]
            
            # Process node
            if combination in deadends:
                continue
                
            if combination in visited:
                continue
                
            if combination == target:
                return num_turns_so_far
            
            visited.add(combination)
            
            # Add neighbors
            neighbors = self.get_neighbors(combination)
            for neighbor_combination in neighbors:
                queue.append({
                    COMBINATION: neighbor_combination,
                    NUM_TURNS_SO_FAR: num_turns_so_far + 1
                })
        
        return IMPOSSIBLE_TO_OPEN
    
    def openLock(self, deadends: List[str], target: str) -> int:
        print(self.go_up('5'))
        
        deadends_set = set(deadends)
        
        if START_COMBINATION in deadends_set or target in deadends_set:
            return IMPOSSIBLE_TO_OPEN
        
        return self.get_min_turns(target, deadends_set)